## **EX NO:**
##   **DATE :**
# <center>**CROSS VALIDATION**</center>

## **AIM:**

 ### To understand and implement cross-validation using python code in Linear Regression to evaluate model performance and prevent overfitting.

## **PROGRAM:**

 **Import Necessary Libraries**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score,KFold
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score

**Load the Dataset**

In [2]:
data = fetch_california_housing()
df = pd.DataFrame(data.data, columns=data.feature_names)
df['Target'] = data.target
# Display first 5 rows
df.head()


NameError: name 'fetch_california_housing' is not defined

**K-Fold Cross-Validation**

In [ ]:
# Define Features and Target Variable
X = df.drop(columns=['Target'])
y = df['Target']
# K-Fold Cross-Validation
# Define model
model = LinearRegression()
# Define K-Fold strategy
kf = KFold(n_splits=5, shuffle=True, random_state=42)
# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')
# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE:", np.mean(cv_rmse))

Cross-Validation RMSE Scores: [0.74558138 0.72639951 0.71364081 0.71052852 0.74510434]
Mean RMSE: 0.728250914247974


 **Evaluating the Model Performance**

In [ ]:
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Train model
model.fit(X_train, y_train)
# Predict and evaluate
y_pred = model.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("Test Set RMSE:", rmse)
# evaluate the model on test set
r2 = r2_score(y_test, y_pred)
print("R square value:",r2)

Test Set RMSE: 0.745581383012776
R square value: 0.5757877060324512


**EXERCISE:**

In [ ]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold, GridSearchCV, LeaveOneOut, cross_val_score, train_test_split
from sklearn.metrics import mean_squared_error, r2_score, make_scorer

**Load the housing dataset**

In [ ]:
df = pd.read_csv("HousingData.csv")
df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93   0.0  0.573  6.030   NaN  2.5050    1  273   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90    NaN  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99    NaN  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

**Data Preprocessing**

In [ ]:
df.isnull().sum()

CRIM       20
ZN         20
INDUS      20
CHAS       20
NOX         0
RM          0
AGE        20
DIS         0
RAD         0
TAX         0
PTRATIO     0
B           0
LSTAT      20
MEDV        0
dtype: int64

In [ ]:
df['CRIM'].fillna(df['CRIM'].mean(),inplace=True)
df['ZN'].fillna(df['ZN'].mean(),inplace=True)
df['INDUS'].fillna(df['INDUS'].mean(),inplace=True)
df['LSTAT'].fillna(df['LSTAT'].mean(),inplace=True)
df['CHAS'].fillna(df['CHAS'].mode()[0],inplace=True)
df['AGE'].fillna(df['AGE'].mode()[0],inplace=True)

**Data after preprocessing**

In [ ]:
df

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575   65.2  4.0900    1  296   
1    0.02731   0.0   7.07   0.0  0.469  6.421   78.9  4.9671    2  242   
2    0.02729   0.0   7.07   0.0  0.469  7.185   61.1  4.9671    2  242   
3    0.03237   0.0   2.18   0.0  0.458  6.998   45.8  6.0622    3  222   
4    0.06905   0.0   2.18   0.0  0.458  7.147   54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...    ...     ...  ...  ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593   69.1  2.4786    1  273   
502  0.04527   0.0  11.93   0.0  0.573  6.120   76.7  2.2875    1  273   
503  0.06076   0.0  11.93   0.0  0.573  6.976   91.0  2.1675    1  273   
504  0.10959   0.0  11.93   0.0  0.573  6.794   89.3  2.3889    1  273   
505  0.04741   0.0  11.93   0.0  0.573  6.030  100.0  2.5050    1  273   

     PTRATIO       B      LSTAT  MEDV  
0       15.3  396.90   4.980000  24.0  
1       17.8  396.90   9.140000  21.6  
2       17.8  392.83   4.030000  34.7  
3       18.7  394.63   2.940000  33.4  
4       18.7  396.90  12.715432  36.2  
..       ...     ...        ...   ...  
501     21.0  391.99  12.715432  22.4  
502     21.0  396.90   9.080000  20.6  
503     21.0  396.90   5.640000  23.9  
504     21.0  393.45   6.480000  22.0  
505     21.0  396.90   7.880000  11.9  

[506 rows x 14 columns]

In [ ]:
df.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

In [ ]:
df.duplicated().sum()

0

**RMSE VALUE BEFORE CROSS-VALIDATION**

In [ ]:
# Building the model
from sklearn.utils import shuffle
df= shuffle(df, random_state=100)
X = df.drop(columns=['MEDV'])
y = df['MEDV']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Build the regression model
import statsmodels.api as sm
X_train_sm = sm.add_constant(X_train)
model = sm.OLS(y_train, X_train_sm).fit()

X_test_sm = sm.add_constant(X_test)
y_pred = model.predict(X_test_sm)

In [ ]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.721
Model:                            OLS   Adj. R-squared:                  0.712
Method:                 Least Squares   F-statistic:                     77.60
Date:                Sun, 16 Feb 2025   Prob (F-statistic):           1.70e-99
Time:                        15:56:12   Log-Likelihood:                -1218.3
No. Observations:                 404   AIC:                             2465.
Df Residuals:                     390   BIC:                             2521.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         31.7543      5.824      5.452      0.000      20.303      43.205
CRIM          -0.1453      0.036     -4.077      0.000      -0.215      -0.075
ZN             0.0400      0.016      2.570      0.011       0.009       0.071
INDUS         -0.0736      0.070     -1.053      0.293      -0.211       0.064
CHAS           2.9579      1.048      2.823      0.005       0.898       5.018
NOX          -15.9214      4.366     -3.647      0.000     -24.505      -7.338
RM             4.2043      0.469      8.965      0.000       3.282       5.126
AGE           -0.0065      0.014     -0.470      0.639      -0.034       0.021
DIS           -1.4827      0.225     -6.586      0.000      -1.925      -1.040
RAD            0.3159      0.076      4.162      0.000       0.167       0.465
TAX           -0.0106      0.004     -2.455      0.015      -0.019      -0.002
PTRATIO       -0.8994      0.155     -5.806      0.000      -1.204      -0.595
B              0.0115      0.003      3.608      0.000       0.005       0.018
LSTAT         -0.4785      0.057     -8.444      0.000      -0.590      -0.367
==============================================================================
Omnibus:                      141.082   Durbin-Watson:                   2.002
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              625.039
Skew:                           1.467   Prob(JB):                    1.88e-136
Kurtosis:                       8.341   Cond. No.                     1.47e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.47e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# Model Validation
X_test_sm = sm.add_constant(X_test)
y_pred = model.predict(X_test_sm)
# Evaluation metrics
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)
print(f"Evaluation Metrics:")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R-squared: {r2:.2f}")

Evaluation Metrics:
Mean Absolute Error (MAE): 3.10
Mean Squared Error (MSE): 16.91
Root Mean Squared Error (RMSE): 4.11
R-squared: 0.76


Before Cross-validation:
  Mean Absolute Error (MAE): 3.10,
  Mean Squared Error (MSE): 16.91,
  Root Mean Squared Error (RMSE): 4.11,
  R-squared: 0.76

#  Implement Stratified K-Fold Cross-Validation and compare results.

In [ ]:
X = df.drop(columns=['MEDV'])  # 'MEDV' is the target
y = df['MEDV']

In [ ]:
model = LinearRegression()
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("K-Fold Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE (K-Fold):", np.mean(cv_rmse))

K-Fold Cross-Validation RMSE Scores: [4.11180286 4.99182163 5.04885871 5.23733036 5.13995332]
Mean RMSE (K-Fold): 4.905953374046761


In [ ]:
# Build the regression model
import statsmodels.api as sm
X_train_sm = sm.add_constant(X_train)
model = sm.OLS(y_train, X_train_sm).fit()

X_test_sm = sm.add_constant(X_test)
y_pred = model.predict(X_test_sm)

In [ ]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.721
Model:                            OLS   Adj. R-squared:                  0.712
Method:                 Least Squares   F-statistic:                     77.60
Date:                Sun, 16 Feb 2025   Prob (F-statistic):           1.70e-99
Time:                        16:00:42   Log-Likelihood:                -1218.3
No. Observations:                 404   AIC:                             2465.
Df Residuals:                     390   BIC:                             2521.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         31.7543      5.824      5.452      0.000      20.303      43.205
CRIM          -0.1453      0.036     -4.077      0.000      -0.215      -0.075
ZN             0.0400      0.016      2.570      0.011       0.009       0.071
INDUS         -0.0736      0.070     -1.053      0.293      -0.211       0.064
CHAS           2.9579      1.048      2.823      0.005       0.898       5.018
NOX          -15.9214      4.366     -3.647      0.000     -24.505      -7.338
RM             4.2043      0.469      8.965      0.000       3.282       5.126
AGE           -0.0065      0.014     -0.470      0.639      -0.034       0.021
DIS           -1.4827      0.225     -6.586      0.000      -1.925      -1.040
RAD            0.3159      0.076      4.162      0.000       0.167       0.465
TAX           -0.0106      0.004     -2.455      0.015      -0.019      -0.002
PTRATIO       -0.8994      0.155     -5.806      0.000      -1.204      -0.595
B              0.0115      0.003      3.608      0.000       0.005       0.018
LSTAT         -0.4785      0.057     -8.444      0.000      -0.590      -0.367
==============================================================================
Omnibus:                      141.082   Durbin-Watson:                   2.002
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              625.039
Skew:                           1.467   Prob(JB):                    1.88e-136
Kurtosis:                       8.341   Cond. No.                     1.47e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.47e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# Stratified K-Fold Cross-Validation
# Bin the target variable (MEDV) into categories for stratification
y_binned = pd.qcut(y, q=5, labels=False)  # Binned into 5 quantiles

# Define the model
model = LinearRegression()

kf = KFold(n_splits=5, shuffle=True, random_state=42)
rmse_scores = []

# Perform Stratified K-Fold Cross-Validation
for train_index, test_index in kf.split(X, y_binned):  # Use binned target for stratification
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]  # Use original target for regression

    # Fit the model
    model.fit(X_train, y_train)

    # Predict on the test set
    y_pred = model.predict(X_test)

    # Calculate RMSE for the fold
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    rmse_scores.append(rmse)

# Output RMSE for each fold and the mean RMSE
print("Stratified K-Fold Cross-Validation RMSE Scores:", rmse_scores)
print("Mean RMSE (Stratified K-Fold):", np.mean(rmse_scores))


Stratified K-Fold Cross-Validation RMSE Scores: [4.111802857652987, 4.991821631670612, 5.048858709124141, 5.237330355700414, 5.1399533160856485]
Mean RMSE (Stratified K-Fold): 4.905953374046761


Before cross validation

In [ ]:
# Train-test split for final evaluation on the test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model on the training set
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Calculate RMSE and R-squared for the test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
test_r2 = r2_score(y_test, y_pred)
print("Test Set RMSE:", test_rmse)
print("R-squared on Test Set:", test_r2)

Test Set RMSE: 4.111802857652993
R-squared on Test Set: 0.762255242412756


**The K-fold and stratified k fold has same mean rmse value of Mean RMSE: 4.9059533740467645 and the rmse value before validation is Test Set RMSE: 4.111802857652992.
Since the test set rmse value is better than both k fold and stratified k fold the model works better with the test set itself.**        

# Experiment with different values of k in K-Fold Cross-Validation.

In [ ]:
# Define Features and Target Variable
X = df.drop(columns=['MEDV'])
y = df['MEDV']

In [ ]:
model = LinearRegression()

For k=10

In [ ]:
kf = KFold(n_splits=10, shuffle=True, random_state=42)

# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("K-Fold Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE (K-Fold):", np.mean(cv_rmse))

K-Fold Cross-Validation RMSE Scores: [3.29867025 4.71204509 3.70288639 6.03541619 4.21529062 6.32821402
 5.41360701 4.81267269 4.36144633 6.00742203]
Mean RMSE (K-Fold): 4.888767062512935


For k=5

In [ ]:
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("K-Fold Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE (K-Fold):", np.mean(cv_rmse))

K-Fold Cross-Validation RMSE Scores: [4.11180286 4.99182163 5.04885871 5.23733036 5.13995332]
Mean RMSE (K-Fold): 4.905953374046761


For k=15

In [ ]:
kf = KFold(n_splits=15, shuffle=True, random_state=42)

# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("K-Fold Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE (K-Fold):", np.mean(cv_rmse))

K-Fold Cross-Validation RMSE Scores: [3.23882937 3.77129024 5.03765046 3.59302159 4.67776889 6.35570101
 4.4860129  6.74565668 4.5207891  5.28495153 5.70209239 4.1058207
 3.47694938 5.4183489  6.18271532]
Mean RMSE (K-Fold): 4.839839898317941


For k=20

In [ ]:
kf = KFold(n_splits=20, shuffle=True, random_state=42)

# Perform Cross-Validation
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse = np.sqrt(-cv_scores)
print("K-Fold Cross-Validation RMSE Scores:", cv_rmse)
print("Mean RMSE (K-Fold):", np.mean(cv_rmse))

K-Fold Cross-Validation RMSE Scores: [2.99822775 4.04340412 4.23875671 4.72278595 3.86905258 3.37860279
 7.75041399 3.72587704 4.96919541 3.21747563 7.88794305 4.6068541
 5.4522846  5.29023532 5.15215698 4.35439282 3.85467545 4.69561015
 4.71692338 6.95637733]
Mean RMSE (K-Fold): 4.794062256033501


**The computed mean rmse values for different k values is in the above code and the mean rmse value of k value 20 is better i.e Mean RMSE (20-Fold): 4.794062256033503 than the k values 5,10,15
So the model performs better when we have done a k=20 fold cross validation**

# Try Leave-One-Out Cross-Validation (LOO-CV) and analyze computation time.

In [ ]:
import time
# Measure computation time for Leave-One-Out Cross-Validation
#Leave-One-Out Cross-Validation
loo = LeaveOneOut()

start_time = time.time()  # Start time for LOO-CV

# Perform LOO-CV
cv_scores_loo = cross_val_score(model, X, y, cv=loo, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse_loo = np.sqrt(-cv_scores_loo)
mean_rmse_loo = np.mean(cv_rmse_loo)

end_time = time.time()  # End time for LOO-CV
loo_time = end_time - start_time  # Compute time taken

print("LOO-CV RMSE Scores:", cv_rmse_loo[:5])  # Display only first 5 scores
print(f"Mean RMSE (LOO-CV): {mean_rmse_loo}")
print(f"LOO-CV Computation Time: {loo_time:.2f} seconds")

LOO-CV RMSE Scores: [0.42735576 0.57398062 1.67962808 3.7469224  4.28260632]
Mean RMSE (LOO-CV): 3.432514822116063
LOO-CV Computation Time: 2.85 seconds


In [ ]:
# Build the regression model
import statsmodels.api as sm
X_train_sm = sm.add_constant(X_train)
model = sm.OLS(y_train, X_train_sm).fit()

X_test_sm = sm.add_constant(X_test)
y_pred = model.predict(X_test_sm)

In [ ]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.721
Model:                            OLS   Adj. R-squared:                  0.712
Method:                 Least Squares   F-statistic:                     77.60
Date:                Sun, 16 Feb 2025   Prob (F-statistic):           1.70e-99
Time:                        16:01:42   Log-Likelihood:                -1218.3
No. Observations:                 404   AIC:                             2465.
Df Residuals:                     390   BIC:                             2521.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         31.7543      5.824      5.452      0.000      20.303      43.205
CRIM          -0.1453      0.036     -4.077      0.000      -0.215      -0.075
ZN             0.0400      0.016      2.570      0.011       0.009       0.071
INDUS         -0.0736      0.070     -1.053      0.293      -0.211       0.064
CHAS           2.9579      1.048      2.823      0.005       0.898       5.018
NOX          -15.9214      4.366     -3.647      0.000     -24.505      -7.338
RM             4.2043      0.469      8.965      0.000       3.282       5.126
AGE           -0.0065      0.014     -0.470      0.639      -0.034       0.021
DIS           -1.4827      0.225     -6.586      0.000      -1.925      -1.040
RAD            0.3159      0.076      4.162      0.000       0.167       0.465
TAX           -0.0106      0.004     -2.455      0.015      -0.019      -0.002
PTRATIO       -0.8994      0.155     -5.806      0.000      -1.204      -0.595
B              0.0115      0.003      3.608      0.000       0.005       0.018
LSTAT         -0.4785      0.057     -8.444      0.000      -0.590      -0.367
==============================================================================
Omnibus:                      141.082   Durbin-Watson:                   2.002
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              625.039
Skew:                           1.467   Prob(JB):                    1.88e-136
Kurtosis:                       8.341   Cond. No.                     1.47e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.47e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
kf = KFold(n_splits=5, shuffle=True, random_state=42)

start_time_kfold = time.time()  # Start time for K-Fold CV

# Perform K-Fold Cross-Validation
cv_scores_kfold = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')

# Convert negative MSE to positive RMSE
cv_rmse_kfold = np.sqrt(-cv_scores_kfold)
mean_rmse_kfold = np.mean(cv_rmse_kfold)

end_time_kfold = time.time()  # End time for K-Fold CV
kfold_time = end_time_kfold - start_time_kfold  # Compute time taken

print("K-Fold RMSE Scores:", cv_rmse_kfold)
print(f"Mean RMSE (K-Fold): {mean_rmse_kfold}")
print(f"K-Fold Computation Time: {kfold_time:.2f} seconds")

K-Fold RMSE Scores: [4.11180286 4.99182163 5.04885871 5.23733036 5.13995332]
Mean RMSE (K-Fold): 4.905953374046761
K-Fold Computation Time: 0.06 seconds


**The mean rmse value of Leave one out cross val is Mean RMSE (LOO-CV): 3.4325148221160626
which is better tha normal k fold,stratified k fold and rmse value of test set which is of 4.9 and 4.11 respectively. So the model performs better when done Leave one out
but the computation time of Leave one out is greater than k fold when compared.**

# Implement Grid Search Function with GridSearchCV for applying different values of k.

In [ ]:
# Split the data into training and test sets (test set for final evaluation)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define the model
model = LinearRegression()

# Define the parameter grid
param_grid = {
    'fit_intercept': [True, False],   # Whether to calculate the intercept
    'positive': [True, False]  # coefficients to be positive
}

# Define the K-Fold strategy within GridSearchCV
k_values = [5, 20, 10, 15,25]  # Different values for the number of folds

# Perform grid search with cross-validation over different values of k
for k in k_values:
    print(f"\nGrid Search with K-Fold (k={k})")

    # Define K-Fold strategy
    kf = KFold(n_splits=k, shuffle=True, random_state=42)

    # Define the scoring metric as RMSE (Root Mean Squared Error)
    scorer = make_scorer(mean_squared_error, greater_is_better=False)

    # Set up GridSearchCV
    grid_search = GridSearchCV(estimator=model,
                               param_grid=param_grid,
                               scoring=scorer,
                               cv=kf,  # K-Fold Cross-Validation with k folds
                               refit=True,  # Refit the best model after search
                               n_jobs=-1)  # Use all available cores

    # Perform the grid search on the training data
    grid_search.fit(X_train, y_train)

    # Get the best model and best parameters
    best_model = grid_search.best_estimator_
    best_params = grid_search.best_params_

    print(f"Best Parameters for k={k}: {best_params}")

    # Evaluate the model on the training set using Cross-Validation RMSE
    best_cv_scores = cross_val_score(best_model, X_train, y_train, cv=kf, scoring='neg_mean_squared_error')
    best_cv_rmse = np.sqrt(-best_cv_scores)

    print(f"Best Cross-Validation RMSE Scores for k={k}: {best_cv_rmse}")
    print(f"Mean RMSE for k={k}: {np.mean(best_cv_rmse)}")


Grid Search with K-Fold (k=5)
Best Parameters for k=5: {'fit_intercept': True, 'positive': False}
Best Cross-Validation RMSE Scores for k=5: [5.09262514 4.233384   5.42941455 5.152361   5.38721952]
Mean RMSE for k=5: 5.059000842533958

Grid Search with K-Fold (k=20)
Best Parameters for k=20: {'fit_intercept': True, 'positive': False}
Best Cross-Validation RMSE Scores for k=20: [5.35548546 3.81340932 6.67213438 3.74153474 5.69472479 3.69674442
 3.94785387 3.56871496 6.99040834 5.30844366 3.76318204 5.72990843
 4.93143551 3.90717407 3.46602859 7.59421629 4.54530762 7.00955245
 3.55809561 6.22575988]
Mean RMSE for k=20: 4.976005721608966

Grid Search with K-Fold (k=10)
Best Parameters for k=10: {'fit_intercept': True, 'positive': False}
Best Cross-Validation RMSE Scores for k=10: [4.69413988 5.49984878 4.63989573 3.80112314 6.24800093 4.80277538
 4.46777441 5.89014412 5.73254754 5.01745418]
Mean RMSE for k=10: 5.079370408497299

Grid Search with K-Fold (k=15)
Best Parameters for k=15: {'

**The value of mean rmse when done grid search function for k values 5,10,15,20,25 the value of k=25 is better which gives a rmse value of Mean RMSE for k=25: 4.951329203191086**

## **TABULATION:**

In [ ]:
pip install prettytable

  Obtaining dependency information for prettytable from https://files.pythonhosted.org/packages/5b/a2/fa0679e7a64b564074a8aa680c664ba8e6770166034f035a22ce74e55ae5/prettytable-3.14.0-py3-none-any.whl.metadata
Note: you may need to restart the kernel to use updated packages.


In [ ]:
# from prettytable import PrettyTable
table = PrettyTable()
table.field_names = ["Stages","RMSE","Mean RMSE"]
table.add_row(["Before cross-validation",4.112,4.11])
table.add_row(["After Stratikied K-Fold(k=5,10,15,20,25)",[4.11, 4.99, 5.04, 5.23, 5.13], 4.905])
table.add_row(["After K-Fold(k=5,10,15,20,25)", [4.905, 4.88, 4.821, 4.83, 4.79],"4.833(Computation time:0.06sec)"])
table.add_row(["After Leave one out",[ 0.427, 0.573, 1.679, 3.74, 4.28], "3.432(Computation time:3.85 sec)"])
table.add_row(["After GridSearchCV(k=5,10,15,20,25)",[4.95,5.01,5.07,4.97,5.05], 4.95])
table.align["Stages"] = "l"
table.align["RMSE"] = "c"
table.align["Mean RMSE"] = "c"
print(table)

+------------------------------------------+-----------------------------------+----------------------------------+
| Stages                                   |                RMSE               |            Mean RMSE             |
+------------------------------------------+-----------------------------------+----------------------------------+
| Before cross-validation                  |               4.112               |               4.11               |
| After Stratikied K-Fold(k=5,10,15,20,25) |   [4.11, 4.99, 5.04, 5.23, 5.13]  |              4.905               |
| After K-Fold(k=5,10,15,20,25)            |  [4.905, 4.88, 4.821, 4.83, 4.79] | 4.833(Computation time:0.06sec)  |
| After Leave one out                      | [0.427, 0.573, 1.679, 3.74, 4.28] | 3.432(Computation time:3.85 sec) |
| After GridSearchCV(k=5,10,15,20,25)      |   [4.95, 5.01, 5.07, 4.97, 5.05]  |               4.95               |
+------------------------------------------+----------------------------

## **RESULT:**

        Thus the program to implement Cross Validation in Linear Regression For the model shows that how much the RMSE value is low the model performs better if the RMSE values after different Cross Validation is low then the type of cross validation is choosen to build a better model and implementing and understanding different type of cross validation such as  K-Fold, Stratified K-Fold, Leave one out and Grid search is executed above successfully.